### PR TITLE
[Merged by Bors] - feat: TwoUniqueProds

### PR DESCRIPTION
**Pull request by alreadydone**
+ Introduce `TwoUniqueProds/Sums`: let `R[G]` be a monoid algebra over a semiring `R` without zero-divisors. A natural sufficient condition for `R[G]` to have no zero-divisors is that `G` has `UniqueProds`, as is shown by [MonoidAlgebra.instNoZeroDivisorsOfUniqueProds](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/MonoidAlgebra/NoZeroDivisors.html#MonoidAlgebra.instNoZeroDivisorsOfUniqueProds). Similarly, a natural sufficient condition for `R[G]` to have only trivial units (of the form `rg` with `r` a unit in `R` and `g` a unit in `G`) is that `G` has `TwoUniqueProds`, but we don't prove this yet in this PR. `TwoUniqueProds G` is also a natural sufficient condition in order for factors of a homogeneous element in an algebra graded by `G` without zero-divisors to themselves be homogeneous.

+ Show `TwoUniqueProds` implies `UniqueProds`: `TwoUniqueProds.toUniqueProds`

+ Strengthen `of_Covariant_right/left` to have `TwoUniqueProds` as conclusion

+ Extract `of_image_filter` from the proof of the instance `UniqueProds (∀ i, G i)` and use it also in the proof of `TwoUniqueProds (∀ i, G i)`

+ Use some private defs (starting from `private abbrev I`) to transfer `(Two)UniqueProds (∀ i, G i)` instances to `(Two)UniqueProds (G × H)`

+ Move the `[Module ℚ G] : UniqueSums G` instance from NoZeroDivisors.lean to UniqueProds.lean and strengthen to `TwoUniqueSums`

New lemmas about UniqueMul:
+ `of_card_le_one`, `iff_card_le_one`, `UniqueMul_of_TwoUniqueMul`

---

- [x] depends on: #6818

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/UniqueProds.lean [hunk 1]
```diff
@@ -184,15 +200,8 @@ theorem of_mulOpposite
...
-  refine of_mulOpposite (G := MulOpposite G) <| fun a b ha hb hab ↦ ?_
-  simp only [mem_map, Function.Embedding.coeFn_mk, exists_exists_and_eq_and] at ha hb
-  rcases ha with ⟨a, ha, rfl⟩
-  rcases hb with ⟨b, hb, rfl⟩
-  rw [op_inj, op_inj, op_inj, op_inj]
-  apply h ha hb ?_
-  apply_fun op ∘ op using op_injective.comp op_injective
-  exact hab
+    UniqueMul (B.map ⟨_, op_injective⟩) (A.map ⟨_, op_injective⟩) (op b0) (op a0) :=
+  of_mulOpposite (by simp_rw [map_map]; exact (mulHom_map_iff _ fun _ _ => by rfl).mpr h)
```
Comment (adomani) [inline Mathlib/Algebra/Group/UniqueProds.lean:204]: Nice golf of my clumsy proof!

File: Mathlib/Algebra/Group/UniqueProds.lean [hunk 2]
```diff
@@ -221,149 +241,141 @@ class UniqueProds (G) [Mul G] : Prop where
...
-    UniqueProds G :=
-⟨fun hA hB =>
-  let f : G ↪ Gᵐᵒᵖ := ⟨op, op_injective⟩
-  let ⟨y, yB, x, xA, hxy⟩ := h.uniqueMul_of_nonempty (hB.map (f := f)) (hA.map (f := f))
-  ⟨unop x, (mem_map' _).mp xA, unop y, (mem_map' _).mp yB, hxy.of_mulOpposite⟩⟩
+private abbrev I : Bool → Type max u v := Bool.rec (ULift.{v} G) (ULift.{u} H)
+@[to_additive] private instance : ∀ b, Mul (I G H b) := Bool.rec ULift.mul ULift.mul
+@[to_additive] private def Prod.upMulHom : G × H →ₙ* ∀ b, I G H b :=
+  ⟨fun x => Bool.rec ⟨x.1⟩ ⟨x.2⟩, fun x y => by ext b; cases b <;> rfl⟩
+@[to_additive] private def downMulHom : ULift G →ₙ* G := ⟨fun x => ULift.down x, fun _ _ => rfl⟩
```
Comment (adomani) [inline Mathlib/Algebra/Group/UniqueProds.lean:315]:
```suggestion
@[to_additive] private def downMulHom : ULift G →ₙ* G := ⟨ULift.down, fun _ _ => rfl⟩
```
Does this work?

File: Mathlib/Algebra/Group/UniqueProds.lean [hunk 3]
```diff
@@ -221,149 +241,141 @@ class UniqueProds (G) [Mul G] : Prop where
...
-    simp_rw [mem_image, mem_filter] at hA hB
-    refine ⟨(aG, aH), ?_, (bG, bH), ?_, fun a b ha hb he => ?_⟩
-    · obtain ⟨a, ⟨ha, rfl⟩, rfl⟩ := hA; exact ha
-    · obtain ⟨b, ⟨hb, rfl⟩, rfl⟩ := hB; exact hb
-    rw [Prod.ext_iff] at he
-    specialize hG (mem_image.mpr ⟨a, ha, rfl⟩) (mem_image.mpr ⟨b, hb, rfl⟩) he.1
-    specialize hH _ _ he.2
-    all_goals try simp_rw [mem_image, mem_filter]
-    exacts [⟨a, ⟨ha, hG.1⟩, rfl⟩, ⟨b, ⟨hb, hG.2⟩, rfl⟩, ⟨Prod.ext hG.1 hH.1, Prod.ext hG.2 hH.2⟩]
+open Finset MulOpposite in
```
Comment (adomani) [inline Mathlib/Algebra/Group/UniqueProds.lean:338]:
```suggestion
open MulOpposite in
```
Is this enough?  There might be an `open Finset` already in scope.

File: Mathlib/Algebra/Group/UniqueProds.lean [hunk 4]
```diff
@@ -221,149 +241,141 @@ class UniqueProds (G) [Mul G] : Prop where
...
+      exacts [ihB B' (hc.2 bi) hB, ihA A' ((A.filter_subset _).ssubset_of_ne hA') hA hB]
     rw [mem_filter] at ha0 hb0
-    refine ⟨a0, ha0.1, b0, hb0.1, fun a b ha hb he => ?_⟩
-    specialize hi (mem_image_of_mem _ ha) (mem_image_of_mem _ hb) ?_
-    · refine (congr_arg (· i) he).trans ?_; rw [← ha0.2, ← hb0.2]; rfl
-    exact hu (mem_filter.mpr ⟨ha, hi.1⟩) (mem_filter.mpr ⟨hb, hi.2⟩) he
+    exact ⟨a0, ha0.1, b0, hb0.1, of_image_filter (Pi.evalMulHom G i) ha0.2 hb0.2 hi hu⟩
+
+open ULift in
+instance [UniqueProds G] [UniqueProds H] : UniqueProds (G × H) := by
```
Comment (adomani) [inline Mathlib/Algebra/Group/UniqueProds.lean:372]:
```suggestion
@[to_additive]
instance [UniqueProds G] [UniqueProds H] : UniqueProds (G × H) := by
```
Is there a reason for not additivizing this?  Have I missed it elsewhere? :)

File: Mathlib/Algebra/Group/UniqueProds.lean [hunk 5]
```diff
@@ -375,3 +387,150 @@ instance {ι} (G : ι → Type*) [∀ i, AddZeroClass (G i)] [∀ i, UniqueSums
...
+        UniqueMul.of_image_filter (Pi.evalMulHom G i) ha2.2 hb2.2 hi2 hu2⟩
+      contrapose! hne; rw [Prod.mk.inj_iff] at hne ⊢
+      rw [← ha1.2, ← hb1.2, ← ha2.2, ← hb2.2, hne.1, hne.2]; exact ⟨rfl, rfl⟩
+    all_goals
+      rcases hc with hc | hc; · exact ihA _ (hc.2 _)
+      by_cases hA : A.filter (· i = _) = A; rw [hA]
+      exacts [ihB _ (hc.2 _), ihA _ ((A.filter_subset _).ssubset_of_ne hA)]
+
+open ULift in
+instance [TwoUniqueProds G] [TwoUniqueProds H] : TwoUniqueProds (G × H) := by
```
Comment (adomani) [inline Mathlib/Algebra/Group/UniqueProds.lean:445]:
```suggestion
@[to_additive]
instance [TwoUniqueProds G] [TwoUniqueProds H] : TwoUniqueProds (G × H) := by
```
Same as above.

File: Mathlib/Algebra/Group/UniqueProds.lean [hunk 6]
```diff
@@ -375,3 +387,150 @@ instance {ι} (G : ι → Type*) [∀ i, AddZeroClass (G i)] [∀ i, UniqueSums
...
+  of_mulOpposite <| (mulHom_image_iff <| MulEquiv.opOp G).mp h
+
+open Finset in
+-- see Note [lower instance priority]
+/-- This instance asserts that if `G` has a right-cancellative multiplication, a linear order, and
+  multiplication is strictly monotone w.r.t. the second argument, then `G` has `TwoUniqueProds`. -/
+@[to_additive
+  "This instance asserts that if `G` has a right-cancellative addition, a linear order,
+  and addition is strictly monotone w.r.t. the second argument, then `G` has `TwoUniqueSums`." ]
+instance (priority := 100) of_Covariant_right [IsRightCancelMul G]
```
Comment (adomani) [inline Mathlib/Algebra/Group/UniqueProds.lean:478]: I was expecting to see `Finset.min'_lt_max'_of_card` somewhere in this proof: does it not help?

**Review by adomani — APPROVED** [review]
Very nice!  I left just a couple of small comments.

File: Mathlib/Algebra/Group/UniqueProds.lean (hunk 6 above)
Comment (alreadydone) [reply Mathlib/Algebra/Group/UniqueProds.lean:480]: Indeed, [the proof using `min'_lt_max'`](https://github.com/leanprover-community/mathlib4/pull/7169/commits/5e10594daffcae239d69b2a3fef847a2ddc21277#diff-ebdf20b185bac5adc5a22ffdf53e7dfa74bb23fb34dfa47a761bb9dfa66c9fc1L494-R501) is shorter and no longer depends on `UniqueMul.iff_card_le_one` whose proof takes a few more lines. Thanks! We still need to use the existence of one unique pair to deduce `1 < (A*B).card` from `1 < (A×ˢB).card` though.

File: Mathlib/Algebra/Group/UniqueProds.lean (hunk 4 above)
Comment (alreadydone) [reply Mathlib/Algebra/Group/UniqueProds.lean]: Thanks for spotting this! Turns out I need a [shortcut instance](https://github.com/leanprover-community/mathlib4/pull/7169/commits/5e10594daffcae239d69b2a3fef847a2ddc21277#diff-ebdf20b185bac5adc5a22ffdf53e7dfa74bb23fb34dfa47a761bb9dfa66c9fc1R313) to make it work.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**